pints/_mcmc/_dual_averaging.py

```diff
@@ -0,0 +1,256 @@
+#
+# Dual Averaging step size and mass matrix adaption method for NUTS and HMC
+# samplers
+#
+# This file is part of PINTS (https://github.com/pints-team/pints/) which is
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+import numpy as np
+
+
+class DualAveragingAdaption:
+    r"""
+    Implements a Dual Averaging scheme to adapt the step size ``epsilon``, as
+    per [1]_ (section 3.2.1 and algorithm 6), and estimates the inverse mass
+    matrix using the sample covariance of the accepted parameter, as suggested
+    in [2]_. The mass matrix can either be given as a fully dense matrix
+    represented as a 2D ndarray, or a diagonal matrix represented as a 1D
+    ndarray.
+
+    During iteration ``m`` of adaption, the parameter ``epsilon`` is updated
+    using the following scheme:
+
+    .. math::
+        \bar{H} = (1 - 1/(m + t_0)) \bar{H} + 1/(m + t_0)(\delta_t - \delta)
+        \text{log} \epsilon = \mu - \sqrt{m}/\gamma \bar{H}
+
+    where $\delta_t$ is the target acceptence probability set by the user and
+    $\delta$ is the acceptence probability reported by the algorithm (i.e. that
+    is provided as an argument to the :meth:`step` method.
+
+    The adaption is done using the same windowing method employed by Stan,
+    which is done over three or more windows:
+
+    - initial window: epsilon is adapted using dual averaging (*no* adaption of
+      the mass matrix).
+    - base window: epsilon continues to be adapted using dual averaging; this
+      adaption completes at the end of this window. The inverse mass matrix is
+      adaped at the end of the window by taking the sample covariance of all
+      parameter points within this window.
+    - terminal window: epsilon is adapted using dual averaging, holding the
+      mass matrix constant, and completes at the end of the window.
+
+    If the number of warmup steps requested by the user is greater than the sum
+    of these three windows, then additional base windows are added, each with a
+    size double that of the previous window.
+
+    References
+    ----------
+    .. [1] Hoffman, M. D., & Gelman, A. (2014). The No-U-Turn sampler:
+           adaptively setting path lengths in Hamiltonian Monte Carlo.
+           Journal of Machine Learning Research, 15(1), 1593-1623.
+
+    .. [2] Betancourt, M. (2018). `A Conceptual Introduction to Hamiltonian
+           Monte Carlo`, https://arxiv.org/abs/1701.02434.
+
+    Attributes
+    ----------
+
```

Comment:
Usually we wouldn't have these blank lines here. Don't particularly care but not sure if it affects the final layout?

Reply:
Also why "attributes" rather than "arguments"? And this list doesn't seem correct?